### PR TITLE
Dataloader update

### PR DESCRIPTION
**Pull request by vhoma**
Add dataloader, nn_architecture and learning files.
Currently runs with just one example and all parameters hardcoded.

### CODE REVIEW

File: data_loader.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        # convert to torch
+        return torch.from_numpy(t1).float(), torch.from_numpy(t2).float(), matrix
+
+    def __len__(self):
+        return len(self.names_array)
+
+
+if __name__ == "__main__":
+    curr_path = "."
+    d_path = "./data"
```
Comment (wf34) [inline data_loader.py]: It is not required to hard-code relative paths here, we can make this a comman- line argument.
`argparse` of standard library enables to do this rather aesthetically:

```
def main(arg_name):
    pass

def get_args():
    parser = argparse.ArgumentParser(description=sys.argv[0])
    parser.add_argument('--arg_name, required=True, type=str, help='ta-da')
    return vars(parser.parse_args())


if '__main__' == __name__:
    main(**get_args())

```

File: data_loader.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    print(f"Unzip '{archive_path}'...")
+    with ZipFile(archive_path, 'r') as zip_obj:
+        zip_obj.extractall(path=data_path)
+
+
+def normalize(img, min_val, max_val):
+    return (img - min_val) / (max_val - min_val)
+
+
+def transform(img):
```
Comment (wf34) [inline data_loader.py]: I would have transform taking in two arguments, the `img` and the `angle`, the angle being randomized elsewhere.
I would add `seed` argument to the tool, to seed the randomizer.
That way we would have consistent results between the experiments.

File: learning.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import copy
+
+import torch
+from torch import nn
+import torch.optim as optim
+from torch.optim import lr_scheduler
+from torch.utils.data import Dataset, DataLoader
+
+
+def train(model, x, y, matrix):
```
Comment (wf34) [inline learning.py]: Still plan to pass to the function a few extra variables:
`num_epochs`, `optimizer`, `scheduler`, `criterion`?

File: learning.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    best_loss = np.inf
+    loss_history = []
+
+    # Iterate throughout the epochs
+    for epoch in range(num_epochs):
```
Comment (wf34) [inline learning.py]: The dataset is virtually unlimited In our case.
So, I don't think its necessary to limit ourselves to the concept of epoch.
That is, to show the network same pair of scans 100 times.
Maybe, we could make little fewer repetitions on each pair, but show more diverse set of pairs, using the same amount of compute?

File: learning.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        # Zero the gradients
+        optimizer.zero_grad()
+        x_gpu = x.to(device)
+        y_gpu = y.to(device)
```
Comment (wf34) [inline learning.py]: These operations actually are not fast. 
Should we from the start adopt more mature architecture, which utilizes GPU better,  and train in **batches**, not in single pairs of scans?

**Review by wf34 — APPROVED** [review]
Good result.
My comments are really **not** a blocker for merge.
At your discretion, you may address a few before merge, and maybe choose to postpone others.

File: data_loader.py (hunk 1 above)
Comment (wf34) [reply data_loader.py]: same goes for the `learning.py`

File: data_loader.py (hunk 2 above)
Comment (vhoma) [reply data_loader.py]: fixed, seed added to cmd args

File: learning.py (hunk 3 above)
Comment (vhoma) [reply learning.py]: vars are now passed to the Learner class

File: learning.py (hunk 4 above)
Comment (vhoma) [reply learning.py]: need to discuss this and add to the next update

File: learning.py (hunk 5 above)
Comment (vhoma) [reply learning.py]: to be fixed in the next update

File: data_loader.py (hunk 1 above)
Comment (vhoma) [reply data_loader.py]: fixed
